libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedModulePath.java

```diff
@@ -0,0 +1,339 @@
+        return builder.build();
+    }
+
+    /** Scans a given path for class files and services. */
+    static ScanResult scan(Path path) throws IOException {
+        Map<Boolean, Set<String>> map = Files.walk(path)
+            .filter(p -> Files.isDirectory(p) == false)
+            .map(p -> path.relativize(p).toString())
+            .filter(p -> (p.endsWith(".class") ^ p.startsWith(SERVICES_PREFIX)))
+            .collect(Collectors.partitioningBy(e -> e.startsWith(SERVICES_PREFIX), Collectors.toSet()));
+        return new ScanResult(map.get(Boolean.FALSE), map.get(Boolean.TRUE));
+    }
+
+    static final Pattern DASH_VERSION = Pattern.compile("-(\\d+(\\.|$))");
+
+    /**
+     * Determines the module version (of an automatic module), given the jar file name. As per,
+     * https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/module/ModuleFinder.html#of(java.nio.file.Path...)
+     */
+    static Optional<ModuleDescriptor.Version> version(String jarName) {
+        if (jarName.endsWith(".jar") == false) {
+            throw new IllegalArgumentException("unexpected jar name: " + jarName);
+        }
+        // drop ".jar"
+        String name = jarName.substring(0, jarName.length() - 4);
+        // find first occurrence of -${NUMBER}. or -${NUMBER}$
+        Matcher matcher = DASH_VERSION.matcher(name);
+        if (matcher.find()) {
+            int start = matcher.start();
+            try {
+                String tail = name.substring(start + 1);
+                return Optional.of(ModuleDescriptor.Version.parse(tail));
+            } catch (IllegalArgumentException ignore) {}
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Parses a set of given service files, and returns a map of service name to list of provider
+     * classes.
+     */
+    static Map<String, List<String>> services(Set<String> serviceFiles, Path path) throws IOException {
+        // map names of service configuration files to service names
+        Set<String> serviceNames = serviceFiles.stream()
+            .map(EmbeddedModulePath::toServiceName)
+            .flatMap(Optional::stream)
+            .collect(Collectors.toSet());
+
+        Map<String, List<String>> map = new HashMap<>();
+        // parse each service configuration file
+        for (String sn : serviceNames) {
+            Path se = path.resolve(SERVICES_PREFIX + sn);
+            List<String> providerClasses = Files.readAllLines(se)
+                .stream()
+                .map(EmbeddedModulePath::dropCommentAndTrim)
+                .filter(Predicate.not(String::isEmpty))
+                .toList();
+            if (providerClasses.isEmpty() == false) {
+                map.put(sn, providerClasses);
+            }
+        }
+        return map;
+    }
+
+    /** Drops comments and trims the given string. */
+    static String dropCommentAndTrim(String line) {
```

Comment:
I understand this is package private so it can be unit tested, but having so many methods available makes it difficult to tell what is the real set of methods that are meant to be used here. Could this and maybe some others be made private and testing just done on the methods intended to be used elsewhere?

Reply:
I moved the code around a bit, and remove the javadoc style comments from "internal" methods. This should help with readability, etc.